File: Clover/app/src/main/java/org/floens/chan/core/repository/ImportExportRepository.java
/*
 * Clover - 4chan browser https://github.com/Floens/Clover/
 * Copyright (C) 2014  Floens
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.floens.chan.core.repository;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;

import com.google.gson.Gson;

import org.floens.chan.core.database.DatabaseHelper;
import org.floens.chan.core.database.DatabaseManager;
import org.floens.chan.core.model.export.ExportedAppSettings;
import org.floens.chan.core.model.export.ExportedBoard;
import org.floens.chan.core.model.export.ExportedFilter;
import org.floens.chan.core.model.export.ExportedLoadable;
import org.floens.chan.core.model.export.ExportedPin;
import org.floens.chan.core.model.export.ExportedPostHide;
import org.floens.chan.core.model.export.ExportedSite;
import org.floens.chan.core.model.orm.Board;
import org.floens.chan.core.model.orm.Filter;
import org.floens.chan.core.model.orm.Loadable;
import org.floens.chan.core.model.orm.Pin;
import org.floens.chan.core.model.orm.SiteModel;
import org.floens.chan.core.model.orm.ThreadHide;
import org.floens.chan.core.settings.ChanSettings;
import org.floens.chan.utils.ByteUtils;
import org.floens.chan.utils.Logger;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.inject.Inject;

public class ImportExportRepository {
    private static final String TAG = "ImportExportRepository";
    private static final byte[] SETTINGS_HEADER_MAGIC_CONST = "CLOVER_SETTINGS".getBytes();
    private static final int SETTINGS_HEADER_MAGIC_CONST_LEN = SETTINGS_HEADER_MAGIC_CONST.length;
    public static final String EXPORT_FILE_NAME = "exported_clover_settings.json";

    // Don't forget to change this when changing any of the Export models.
    // Also, don't forget to handle the change in the onUpgrade method.
    public static final int CURRENT_EXPORT_SETTINGS_VERSION = 2;

    private DatabaseManager databaseManager;
    private DatabaseHelper databaseHelper;
    private Gson gson;

    private AtomicBoolean exportingRunning = new AtomicBoolean(false);
    private AtomicBoolean importingRunning = new AtomicBoolean(false);

    @Inject
    public ImportExportRepository(
            DatabaseManager databaseManager,
            DatabaseHelper databaseHelper,
            Gson gson
    ) {
        this.databaseManager = databaseManager;
        this.databaseHelper = databaseHelper;
        this.gson = gson;
    }

    public void exportTo(File settingsFile, ExportCallbacks callbacks) {
        if (!exportingRunning.compareAndSet(false, true)) {
            return;
        }

        databaseManager.runTaskAsync(() -> {
            try {
                ExportedAppSettings appSettings = readSettingsFromDatabase();
                if (appSettings.isEmpty()) {
                    callbacks.onNothingToExport();
                    return null;
                }

                String json = gson.toJson(appSettings);

                if (settingsFile.exists()) {
                    if (!settingsFile.isFile()) {
                        throw new IOException(
                                "Settings file is not a file (???) " + settingsFile.getAbsolutePath()
                        );
                    }

                    if (!settingsFile.delete()) {
                        Logger.w(TAG, "Could not delete export file before exporting " + settingsFile.getAbsolutePath());
                    }
                }

                if (!settingsFile.createNewFile()) {
                    throw new IOException(
                            "Could not create a file for exporting " + settingsFile.getAbsolutePath()
                    );
                }

                if (!settingsFile.exists() || !settingsFile.canWrite()) {
                    throw new IOException(
                            "Something wrong with export file (Can't write or it doesn't exist) "
                                    + settingsFile.getAbsolutePath()
                    );
                }

                try (RandomAccessFile raf = new RandomAccessFile(settingsFile, "rw")) {
                    // Write the header magic constant so we can figure out that this file is the settings file
                    raf.write(SETTINGS_HEADER_MAGIC_CONST);
                    // Write the version of the file
                    raf.writeInt(ImportExportRepository.CURRENT_EXPORT_SETTINGS_VERSION);
                    // Write the settings
                    raf.writeUTF(json);
                }

                Logger.d(TAG, "Exporting done!");
                callbacks.onSuccessExport();
            } catch (Throwable error) {
                Logger.e(TAG, "Error while trying to export settings", error);

                deleteExportFile(settingsFile);
                callbacks.onExportError(error);

                // throw to upstream to rollback the transaction
                throw error;
            } finally {
                exportingRunning.set(false);
            }

            return null;
        });
    }

    public void importFrom(File settingsFile, ImportCallbacks callbacks) {
        if (!importingRunning.compareAndSet(false, true)) {
            return;
        }

        databaseManager.runTaskAsync(() -> {
            try {
                if (!settingsFile.exists()) {
                    Logger.d(TAG, "There is nothing to import, importFile does not exist " +
                            settingsFile.getAbsolutePath());

                    callbacks.onNothingToImport();
                    return null;
                }

                if (!settingsFile.canRead()) {
                    throw new IOException(
                            "Something wrong with import file (Can't read or it doesn't exist) "
                                    + settingsFile.getAbsolutePath()
                    );
                }

                ExportedAppSettings appSettings;
                int settingsFileVersion;

                try (RandomAccessFile raf = new RandomAccessFile(settingsFile, "r")) {
                    byte[] readMagicConst = new byte[SETTINGS_HEADER_MAGIC_CONST_LEN];
                    int readCount = raf.read(readMagicConst);

                    if (readCount != SETTINGS_HEADER_MAGIC_CONST_LEN) {
                        throw new IOException("Could not read settings header magic constant (the " +
                                "amount of bytes read is not equals to the header const length: " + readCount + ")");
                    }

                    if (!Arrays.equals(readMagicConst, SETTINGS_HEADER_MAGIC_CONST)) {
                        throw new IOException("The file does not contain the settings header magic constant, " +
                                "file header is: " + ByteUtils.bytesToHex(readMagicConst) +
                                ", but should be: " + ByteUtils.bytesToHex(SETTINGS_HEADER_MAGIC_CONST));
                    }

                    settingsFileVersion = raf.readInt();
                    Logger.d(TAG, "Settings file version = " + settingsFileVersion);

                    throwIfDowngrade(settingsFileVersion);
                    throwIfCannotUpgrade(settingsFileVersion);

                    appSettings = gson.fromJson(raf.readUTF(), ExportedAppSettings.class);
                }

                if (appSettings.isEmpty()) {
                    Logger.d(TAG, "There is nothing to import, appSettings is empty");
                    callbacks.onNothingToImport();
                    return null;
                }

                writeSettingsToDatabase(settingsFileVersion, appSettings);

                Logger.d(TAG, "Importing success!");
                callbacks.onSuccessImport();
            } catch (Throwable error) {
                Logger.e(TAG, "Error while trying to import settings", error);
                callbacks.onImportError(error);

                // throw to upstream to rollback the transaction
                throw error;
            } finally {
                importingRunning.set(false);
            }

            return null;
        });
    }

    private void throwIfCannotUpgrade(int settingsFileVersion) throws UpgradeImpossibleException {
        if (settingsFileVersion == 1) {
            ///////////////////////////////////////
            // Upgrade from version 1.
            // In version 2 we had to change the settings file format (Version code is moved out of the json part to the header part).
            ///////////////////////////////////////
            throw new UpgradeImpossibleException("Upgrade from version 1 to version 2 is impossible because " +
                    "in version 2 the settings file format had to be changed because of unforeseen problems");
        }
    }

    private void throwIfDowngrade(int settingsFileVersion) throws DowngradeNotSupportedException {
        if (settingsFileVersion > CURRENT_EXPORT_SETTINGS_VERSION) {
            // we don't support settings downgrade so just notify the user about it
            throw new DowngradeNotSupportedException(
                    "You are attempting to import settings with version " +
                            "higher than the current app's settings version (downgrade). " +
                            "This is not supported so nothing will be imported. " +
                            "(Current version = " + ImportExportRepository.CURRENT_EXPORT_SETTINGS_VERSION +
                            ", settingFileVersion = " + settingsFileVersion
            );
        }
    }

    private void deleteExportFile(File exportFile) {
        if (exportFile != null) {
            if (!exportFile.delete()) {
                Logger.w(TAG, "Could not delete export file " + exportFile.getAbsolutePath());
            }
        }
    }

    private void writeSettingsToDatabase(
            int settingsFileVersion,
            @NonNull ExportedAppSettings appSettingsParam)
            throws SQLException, IOException {
        ExportedAppSettings appSettings = appSettingsParam;

        if (settingsFileVersion < CURRENT_EXPORT_SETTINGS_VERSION) {
            appSettings = onUpgrade(settingsFileVersion, appSettings);
        }

        databaseHelper.siteDao.deleteBuilder().delete();
        databaseHelper.loadableDao.deleteBuilder().delete();
        databaseHelper.pinDao.deleteBuilder().delete();
        databaseHelper.boardsDao.deleteBuilder().delete();
        databaseHelper.filterDao.deleteBuilder().delete();
        databaseHelper.threadHideDao.deleteBuilder().delete();

        for (ExportedBoard exportedBoard : appSettings.getExportedBoards()) {
            databaseHelper.boardsDao.createIfNotExists(new Board(
                    exportedBoard.getSiteId(),
                    exportedBoard.isSaved(),
                    exportedBoard.getOrder(),
                    exportedBoard.getName(),
                    exportedBoard.getCode(),
                    exportedBoard.isWorkSafe(),
                    exportedBoard.getPerPage(),
                    exportedBoard.getPages(),
                    exportedBoard.getMaxFileSize(),
                    exportedBoard.getMaxWebmSize(),
                    exportedBoard.getMaxCommentChars(),
                    exportedBoard.getBumpLimit(),
                    exportedBoard.getImageLimit(),
                    exportedBoard.getCooldownThreads(),
                    exportedBoard.getCooldownReplies(),
                    exportedBoard.getCooldownImages(),
                    exportedBoard.getCooldownRepliesIntra(),
                    exportedBoard.getCooldownImagesIntra(),
                    exportedBoard.isSpoilers(),
                    exportedBoard.getCustomSpoilers(),
                    exportedBoard.isUserIds(),
                    exportedBoard.isCodeTags(),
                    exportedBoard.isPreuploadCaptcha(),
                    exportedBoard.isCountryFlags(),
                    exportedBoard.isTrollFlags(),
                    exportedBoard.isMathTags(),
                    exportedBoard.getDescription(),
                    exportedBoard.isArchive()
            ));
        }

        for (ExportedSite exportedSite : appSettings.getExportedSites()) {
            SiteModel inserted = databaseHelper.siteDao.createIfNotExists(new SiteModel(
                    exportedSite.getSiteId(),
                    exportedSite.getConfiguration(),
                    exportedSite.getUserSettings(),
                    exportedSite.getOrder()
            ));

            for (ExportedPin exportedPin : exportedSite.getExportedPins()) {
                Loadable loadable = Loadable.importLoadable(
                        inserted.id,
                        exportedPin.getExportedLoadable().getMode(),
                        exportedPin.getExportedLoadable().getBoardCode(),
                        exportedPin.getExportedLoadable().getNo(),
                        exportedPin.getExportedLoadable().getTitle(),
                        exportedPin.getExportedLoadable().getListViewIndex(),
                        exportedPin.getExportedLoadable().getListViewTop(),
                        exportedPin.getExportedLoadable().getLastViewed(),
                        exportedPin.getExportedLoadable().getLastLoaded()
                );
                databaseHelper.loadableDao.createIfNotExists(loadable);

                Pin pin = new Pin(
                        loadable,
                        exportedPin.isWatching(),
                        exportedPin.getWatchLastCount(),
                        exportedPin.getWatchNewCount(),
                        exportedPin.getQuoteLastCount(),
                        exportedPin.getQuoteNewCount(),
                        exportedPin.isError(),
                        exportedPin.getThumbnailUrl(),
                        exportedPin.getOrder(),
                        exportedPin.isArchived()
                );
                databaseHelper.pinDao.createIfNotExists(pin);
            }
        }

        for (ExportedFilter exportedFilter : appSettings.getExportedFilters()) {
            databaseHelper.filterDao.createIfNotExists(new Filter(
                    exportedFilter.isEnabled(),
                    exportedFilter.getType(),
                    exportedFilter.getPattern(),
                    exportedFilter.isAllBoards(),
                    exportedFilter.getBoards(),
                    exportedFilter.getAction(),
                    exportedFilter.getColor()
            ));
        }

        for (ExportedPostHide exportedPostHide : appSettings.getExportedPostHides()) {
            databaseHelper.threadHideDao.createIfNotExists(new ThreadHide(
                    exportedPostHide.getSite(),
                    exportedPostHide.getBoard(),
                    exportedPostHide.getNo()
            ));
        }

        ChanSettings.deserializeFromString(appSettingsParam.getSettings());
    }

    @NonNull
    private ExportedAppSettings onUpgrade(
            int settingsFileVersion,
            ExportedAppSettings appSettings) {
        // Add your ExportAppSettings migrations here
        return appSettings;
    }

    @NonNull
    private ExportedAppSettings readSettingsFromDatabase() throws java.sql.SQLException, IOException {
        @SuppressLint("UseSparseArrays")
        Map<Integer, SiteModel> sitesMap = new HashMap<>();
        {
            List<SiteModel> sites = databaseHelper.siteDao.queryForAll();

            for (SiteModel site : sites) {
                sitesMap.put(site.id, site);
            }
        }

        @SuppressLint("UseSparseArrays")
        Map<Integer, Loadable> loadableMap = new HashMap<>();
        {
            List<Loadable> loadables = databaseHelper.loadableDao.queryForAll();

            for (Loadable loadable : loadables) {
                loadableMap.put(loadable.id, loadable);
            }
        }

        Set<Pin> pins = new HashSet<>(databaseHelper.pinDao.queryForAll());
        Map<SiteModel, List<ExportedPin>> toExportMap = new HashMap<>();

        for (SiteModel siteModel : sitesMap.values()) {
            toExportMap.put(siteModel, new ArrayList<>());
        }

        for (Pin pin : pins) {
            Loadable loadable = loadableMap.get(pin.loadable.id);
            if (loadable == null) {
                throw new NullPointerException("Could not find Loadable by pin.loadable.id " + pin.loadable.id);
            }

            SiteModel siteModel = sitesMap.get(loadable.siteId);
            if (siteModel == null) {
                throw new NullPointerException("Could not find siteModel by loadable.siteId " + loadable.siteId);
            }

            ExportedLoadable exportedLoadable = new ExportedLoadable(
                    loadable.boardCode,
                    loadable.id,
                    loadable.lastLoaded,
                    loadable.lastViewed,
                    loadable.listViewIndex,
                    loadable.listViewTop,
                    loadable.mode,
                    loadable.no,
                    loadable.siteId,
                    loadable.title
            );

            ExportedPin exportedPin = new ExportedPin(
                    pin.archived,
                    pin.id,
                    pin.isError,
                    pin.loadable.id,
                    pin.order,
                    pin.quoteLastCount,
                    pin.quoteNewCount,
                    pin.thumbnailUrl,
                    pin.watchLastCount,
                    pin.watchNewCount,
                    pin.watching,
                    exportedLoadable
            );

            toExportMap.get(siteModel).add(exportedPin);
        }

        List<ExportedSite> exportedSites = new ArrayList<>();

        for (Map.Entry<SiteModel, List<ExportedPin>> entry : toExportMap.entrySet()) {
            ExportedSite exportedSite = new ExportedSite(
                    entry.getKey().id,
                    entry.getKey().configuration,
                    entry.getKey().order,
                    entry.getKey().userSettings,
                    entry.getValue()
            );

            exportedSites.add(exportedSite);
        }

        List<ExportedBoard> exportedBoards = new ArrayList<>();

        for (Board board : databaseHelper.boardsDao.queryForAll()) {
            exportedBoards.add(new ExportedBoard(
                    board.siteId,
                    board.saved,
                    board.order,
                    board.name,
                    board.code,
                    board.workSafe,
                    board.perPage,
                    board.pages,
                    board.maxFileSize,
                    board.maxWebmSize,
                    board.maxCommentChars,
                    board.bumpLimit,
                    board.imageLimit,
                    board.cooldownThreads,
                    board.cooldownReplies,
                    board.cooldownImages,
                    board.cooldownRepliesIntra,
                    board.cooldownImagesIntra,
                    board.spoilers,
                    board.customSpoilers,
                    board.userIds,
                    board.codeTags,
                    board.preuploadCaptcha,
                    board.countryFlags,
                    board.trollFlags,
                    board.mathTags,
                    board.description,
                    board.archive
            ));
        }

        List<ExportedFilter> exportedFilters = new ArrayList<>();

        for (Filter filter : databaseHelper.filterDao.queryForAll()) {
            exportedFilters.add(new ExportedFilter(
                    filter.enabled,
                    filter.type,
                    filter.pattern,
                    filter.allBoards,
                    filter.boards,
                    filter.action,
                    filter.color
            ));
        }

        List<ExportedPostHide> exportedPostHides = new ArrayList<>();

        for (ThreadHide threadHide : databaseHelper.threadHideDao.queryForAll()) {
            exportedPostHides.add(new ExportedPostHide(
                    threadHide.site,
                    threadHide.board,
                    threadHide.no
            ));
        }

        String settings = ChanSettings.serializeToString();

        return new ExportedAppSettings(
                exportedSites,
                exportedBoards,
                exportedFilters,
                exportedPostHides,
                settings
        );
    }

    public interface ImportCallbacks {
        void onSuccessImport();

        void onNothingToImport();

        void onImportError(Throwable error);
    }

    public interface ExportCallbacks {
        void onSuccessExport();

        void onNothingToExport();

        void onExportError(Throwable error);
    }

    public static class DowngradeNotSupportedException extends Exception {
        public DowngradeNotSupportedException(String message) {
            super(message);
        }
    }

    public static class UpgradeImpossibleException extends Exception {
        public UpgradeImpossibleException(String message) {
            super(message);
        }
    }
}

File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/repository/ImportExportRepository.kt
/*
 * Kuroba - *chan browser https://github.com/Adamantcheese/Kuroba/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.core.repository

import com.github.adamantcheese.chan.core.database.DatabaseHelper
import com.github.adamantcheese.chan.core.database.DatabaseUtils
import com.github.adamantcheese.chan.core.di.AppModule
import com.github.adamantcheese.chan.core.model.export.*
import com.github.adamantcheese.chan.core.model.orm.*
import com.github.adamantcheese.chan.core.repository.ImportExportRepository.ImportExport.Export
import com.github.adamantcheese.chan.core.repository.ImportExportRepository.ImportExport.Import
import com.github.adamantcheese.chan.core.settings.ChanSettings
import com.github.adamantcheese.chan.core.settings.ChanSettings.EMPTY_JSON
import com.github.adamantcheese.chan.utils.Logger
import com.github.k1rakishou.fsaf.FileManager
import com.github.k1rakishou.fsaf.file.AbstractFile
import com.github.k1rakishou.fsaf.file.ExternalFile
import com.github.k1rakishou.fsaf.file.FileDescriptorMode
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.sql.SQLException
import java.util.*
import java.util.regex.Pattern

class ImportExportRepository
constructor(
        private val databaseHelper: DatabaseHelper,
        private val fileManager: FileManager
) {

    fun exportTo(settingsFile: ExternalFile, isNewFile: Boolean, callbacks: ImportExportCallbacks) {
        DatabaseUtils.runTask {
            try {
                val appSettings = readSettingsFromDatabase()
                if (appSettings.isEmpty) {
                    callbacks.onNothingToImportExport(Export)
                    return@runTask
                }

                val json = AppModule.gson.toJson(appSettings)

                if (!fileManager.exists(settingsFile) || !fileManager.canWrite(settingsFile)) {
                    throw IOException(
                            "Something wrong with export file (Can't write or it doesn't exist) "
                                    + settingsFile.getFullPath()
                    )
                }

                // If the user has opened an old settings file we need to use WriteTruncate mode
                // so that there no leftovers of the old file after writing the settings.
                // Otherwise use Write mode
                var fdm = FileDescriptorMode.WriteTruncate
                if (isNewFile) {
                    fdm = FileDescriptorMode.Write
                }

                fileManager.withFileDescriptor(settingsFile, fdm) { fileDescriptor ->
                    FileWriter(fileDescriptor).use { writer ->
                        writer.write(json)
                        writer.flush()
                    }

                    Logger.d(TAG, "Exporting done!")
                    callbacks.onSuccess(Export)
                }

            } catch (error: Throwable) {
                Logger.e(TAG, "Error while trying to export settings", error)

                deleteExportFile(settingsFile)
                callbacks.onError(error, Export)
            }
        }
    }

    fun importFrom(settingsFile: ExternalFile, callbacks: ImportExportCallbacks) {
        DatabaseUtils.runTask {
            try {
                if (!fileManager.exists(settingsFile)) {
                    Logger.e(TAG, "There is nothing to import, importFile does not exist "
                            + settingsFile.getFullPath())
                    callbacks.onNothingToImportExport(Import)
                    return@runTask
                }

                if (!fileManager.canRead(settingsFile)) {
                    throw IOException(
                            "Something wrong with import file (Can't read or it doesn't exist) "
                                    + settingsFile.getFullPath()
                    )
                }

                fileManager.withFileDescriptor(
                        settingsFile,
                        FileDescriptorMode.Read
                ) { fileDescriptor ->
                    FileReader(fileDescriptor).use { reader ->
                        val appSettings = AppModule.gson.fromJson(reader, ExportedAppSettings::class.java)

                        if (appSettings.isEmpty) {
                            Logger.d(TAG, "There is nothing to import, appSettings is empty")
                            callbacks.onNothingToImportExport(Import)
                            return@use
                        }

                        writeSettingsToDatabase(appSettings)

                        Logger.d(TAG, "Importing done!")
                        callbacks.onSuccess(Import)
                    }
                }

            } catch (error: Throwable) {
                Logger.e(TAG, "Error while trying to import settings", error)
                callbacks.onError(error, Import)
            }
        }
    }

    private fun deleteExportFile(exportFile: AbstractFile) {
        if (!fileManager.delete(exportFile)) {
            Logger.w(TAG, "Could not delete export file " + exportFile.getFullPath())
        }
    }

    @Throws(SQLException::class, IOException::class, DowngradeNotSupportedException::class)
    private fun writeSettingsToDatabase(appSettingsParam: ExportedAppSettings) {
        var appSettings = appSettingsParam

        if (appSettings.version < CURRENT_EXPORT_SETTINGS_VERSION) {
            appSettings = onUpgrade(appSettings.version, appSettings)
        } else if (appSettings.version > CURRENT_EXPORT_SETTINGS_VERSION) {
            // we don't support settings downgrade so just notify the user about it
            throw DowngradeNotSupportedException("You are attempting to import settings with " +
                    "version higher than the current app's settings version (downgrade). " +
                    "This is not supported so nothing will be imported."
            )
        }

        // recreate tables from scratch, because we need to reset database IDs as well
        databaseHelper.connectionSource.use { cs ->
            databaseHelper.dropTables(cs)
            databaseHelper.createTables(cs)
        }

        for (exportedBoard in appSettings.exportedBoards) {
            databaseHelper.boardDao.createIfNotExists(Board(
                    exportedBoard.siteId,
                    exportedBoard.isSaved,
                    exportedBoard.order,
                    exportedBoard.name,
                    exportedBoard.code,
                    exportedBoard.isWorkSafe,
                    exportedBoard.perPage,
                    exportedBoard.pages,
                    exportedBoard.maxFileSize,
                    exportedBoard.maxWebmSize,
                    exportedBoard.maxCommentChars,
                    exportedBoard.bumpLimit,
                    exportedBoard.imageLimit,
                    exportedBoard.cooldownThreads,
                    exportedBoard.cooldownReplies,
                    exportedBoard.cooldownImages,
                    exportedBoard.isSpoilers,
                    exportedBoard.customSpoilers,
                    exportedBoard.isUserIds,
                    exportedBoard.isCodeTags,
                    exportedBoard.isPreuploadCaptcha,
                    exportedBoard.isCountryFlags,
                    exportedBoard.boardFlags,
                    exportedBoard.isMathTags,
                    exportedBoard.description ?: "",
                    exportedBoard.isArchive
            ))
        }

        for (exportedSite in appSettings.exportedSites) {
            val inserted = databaseHelper.siteModelDao.createIfNotExists(SiteModel(
                    exportedSite.siteId,
                    exportedSite.configuration,
                    exportedSite.userSettings,
                    exportedSite.order,
                    exportedSite.classId
            ))

            for (exportedPin in exportedSite.exportedPins) {
                val exportedLoadable = exportedPin.exportedLoadable ?: continue

                val loadable = Loadable.importLoadable(
                        inserted.id,
                        exportedLoadable.mode,
                        exportedLoadable.boardCode,
                        exportedLoadable.no,
                        exportedLoadable.title,
                        exportedLoadable.listViewIndex,
                        exportedLoadable.listViewTop,
                        exportedLoadable.lastViewed,
                        exportedLoadable.lastLoaded
                )

                val insertedLoadable = databaseHelper.loadableDao.createIfNotExists(loadable)
                val pin = Pin(
                        insertedLoadable,
                        exportedPin.isWatching,
                        exportedPin.watchLastCount,
                        exportedPin.watchNewCount,
                        exportedPin.quoteLastCount,
                        exportedPin.quoteNewCount,
                        exportedPin.isError,
                        exportedPin.order,
                        exportedPin.isArchived
                )
                databaseHelper.pinDao.createIfNotExists(pin)
            }
        }

        for (exportedFilter in appSettings.exportedFilters) {
            databaseHelper.filterDao.createIfNotExists(Filter(
                    exportedFilter.isEnabled,
                    exportedFilter.type,
                    exportedFilter.pattern,
                    exportedFilter.isAllBoards,
                    exportedFilter.boards,
                    exportedFilter.action,
                    exportedFilter.color,
                    exportedFilter.applyToReplies,
                    exportedFilter.order,
                    exportedFilter.onlyOnOP,
                    exportedFilter.applyToSaved
            ))
        }

        for (exportedPostHide in appSettings.exportedPostHides) {
            databaseHelper.postHideDao.createIfNotExists(PostHide(
                    exportedPostHide.site,
                    exportedPostHide.board,
                    exportedPostHide.no))
        }

        ChanSettings.deserializeFromString(appSettingsParam.settings)
    }

    private fun onUpgrade(version: Int, appSettings: ExportedAppSettings): ExportedAppSettings {
        if (version < 2) {
            //clear the post hides for version 1, threadNo field was added
            appSettings.exportedPostHides = ArrayList()
        }

        if (version < 3) {
            //clear the site model usersettings to be an empty JSON map for version 2,
            // as they won't parse correctly otherwise
            for (site in appSettings.exportedSites) {
                site.userSettings = EMPTY_JSON
            }
        }

        //can't directly use gson here, gotta use regex instead
        //I don't know why, but for some reason Android fails to compile this without the redundant escape??
        @Suppress("RegExpRedundantEscape") val oldConfigPattern = Pattern.compile("\\{\"internal_site_id\":(\\d+),\"external\":.+\\}")

        if (version < 4) {
            //55chan and 8chan were removed for this version
            var chan8: ExportedSite? = null
            var chan55: ExportedSite? = null

            for (site in appSettings.exportedSites) {
                val matcher = oldConfigPattern.matcher(site.configuration.toString())
                if (matcher.matches()) {
                    val classID = matcher.group(1)?.let { Integer.parseInt(it) }
                    if (classID == 1 && chan8 == null) {
                        chan8 = site
                    }

                    if (classID == 7 && chan55 == null) {
                        chan55 = site
                    }
                }
            }

            if (chan55 != null) {
                deleteExportedSite(chan55, appSettings)
            }

            if (chan8 != null) {
                deleteExportedSite(chan8, appSettings)
            }
        }

        if (version < 5) {
            // siteconfig class removed, move stuff over
            for (site in appSettings.exportedSites) {
                val matcher = oldConfigPattern.matcher(site.configuration.toString())
                if (matcher.matches()) {
                    val classID = matcher.group(1)?.let { Integer.parseInt(it) }
                    if (classID != null) {
                        site.classId = classID
                    }
                }
            }
        }

        if (version < 6) {
            for (board in appSettings.exportedBoards) {
                board.boardFlags = HashMap()
            }
        }
        return appSettings
    }

    @Throws(java.sql.SQLException::class, IOException::class)
    private fun readSettingsFromDatabase(): ExportedAppSettings {
        val sitesMap = fillSitesMap()
        val loadableMap = fillLoadablesMap()

        val pins = HashSet(databaseHelper.pinDao.queryForAll())
        val toExportMap = HashMap<SiteModel, MutableList<ExportedPin>>()

        for (siteModel in sitesMap.values) {
            toExportMap[siteModel] = ArrayList()
        }

        for (pin in pins) {
            val loadable = loadableMap[pin.loadable.id]
                    ?: throw NullPointerException("Could not find Loadable by pin.loadable.id "
                            + pin.loadable.id)

            val siteModel = sitesMap[loadable.siteId]
                    ?: throw NullPointerException("Could not find siteModel by loadable.siteId "
                            + loadable.siteId)

            val exportedLoadable = ExportedLoadable(
                    loadable.boardCode,
                    loadable.id.toLong(),
                    loadable.lastLoaded,
                    loadable.lastViewed,
                    loadable.listViewIndex,
                    loadable.listViewTop,
                    loadable.mode,
                    loadable.no,
                    loadable.siteId,
                    loadable.title,
                    loadable.thumbnailUrl?.toString()
            )

            val exportedPin = ExportedPin(
                    pin.archived,
                    pin.id,
                    pin.isError,
                    loadable.id,
                    pin.order,
                    pin.quoteLastCount,
                    pin.quoteNewCount,
                    pin.watchLastCount,
                    pin.watchNewCount,
                    pin.watching,
                    exportedLoadable
            )

            toExportMap[siteModel]!!.add(exportedPin)
        }

        val exportedSites = ArrayList<ExportedSite>()

        for ((key, value) in toExportMap) {
            val exportedSite = ExportedSite(
                    key.id,
                    key.configuration,
                    key.order,
                    key.userSettings,
                    key.classID,
                    value
            )

            exportedSites.add(exportedSite)
        }

        val exportedBoards = databaseHelper.boardDao.queryForAll().map {
            ExportedBoard(
                    it.siteId,
                    it.saved,
                    it.order,
                    it.name,
                    it.code,
                    it.workSafe,
                    it.perPage,
                    it.pages,
                    it.maxFileSize,
                    it.maxWebmSize,
                    it.maxCommentChars,
                    it.bumpLimit,
                    it.imageLimit,
                    it.cooldownThreads,
                    it.cooldownReplies,
                    it.cooldownImages,
                    it.spoilers,
                    it.customSpoilers,
                    it.userIds,
                    it.codeTags,
                    it.preuploadCaptcha,
                    it.countryFlags,
                    it.boardFlags,
                    it.mathTags,
                    it.description,
                    it.archive
            )
        }

        val exportedFilters = databaseHelper.filterDao.queryForAll().map {
            ExportedFilter(
                    it.enabled,
                    it.type,
                    it.pattern,
                    it.allBoards,
                    it.boards,
                    it.action,
                    it.color,
                    it.applyToReplies,
                    it.order,
                    it.onlyOnOP,
                    it.applyToSaved
            )
        }

        val exportedPostHides = databaseHelper.postHideDao.queryForAll().map {
            ExportedPostHide(
                    it.site,
                    it.board,
                    it.no,
                    it.wholeThread,
                    it.hide,
                    it.hideRepliesToThisPost,
                    it.threadNo
            )
        }

        val settings = ChanSettings.serializeToString()

        return ExportedAppSettings(CURRENT_EXPORT_SETTINGS_VERSION,
                exportedSites,
                exportedBoards,
                exportedFilters,
                exportedPostHides,
                settings
        )
    }

    private fun fillLoadablesMap(): Map<Int, Loadable> {
        val map = hashMapOf<Int, Loadable>()
        val loadables = databaseHelper.loadableDao.queryForAll()

        for (loadable in loadables) {
            map[loadable.id] = loadable
        }

        return map
    }

    private fun fillSitesMap(): Map<Int, SiteModel> {
        val map = hashMapOf<Int, SiteModel>()
        val sites = databaseHelper.siteModelDao.queryForAll()

        for (site in sites) {
            map[site.id] = site
        }

        return map
    }

    private fun deleteExportedSite(site: ExportedSite, appSettings: ExportedAppSettings) {
        //filters
        val filtersToDelete = ArrayList<ExportedFilter>()
        for (filter in appSettings.exportedFilters) {
            if (filter.isAllBoards || filter.boards.isNullOrEmpty()) {
                continue
            }

            val boards = checkNotNull(filter.boards)
            val splitBoards = boards.split(",".toRegex()).dropLastWhile { it.isEmpty() }

            for (uniqueId in splitBoards) {
                val split = uniqueId.split(":".toRegex()).dropLastWhile { it.isEmpty() }

                if (split.size == 2 && Integer.parseInt(split[0]) == site.siteId) {
                    filtersToDelete.add(filter)
                    break
                }
            }
        }

        appSettings.exportedFilters.removeAll(filtersToDelete)

        //boards
        val boardsToDelete = appSettings.exportedBoards.filter { it.siteId == site.siteId }
        appSettings.exportedBoards.removeAll(boardsToDelete)

        //loadables for saved threads
        val loadables = ArrayList<ExportedLoadable>()
        for (pin in site.exportedPins) {
            val loadable = pin.exportedLoadable
                    ?: continue

            if (loadable.siteId == site.siteId) {
                loadables.add(loadable)
            }
        }

        //post hides
        val hidesToDelete = appSettings.exportedPostHides.filter { it.site == site.siteId }

        appSettings.exportedPostHides.removeAll(hidesToDelete)

        //site (also removes pins and loadables)
        appSettings.exportedSites.remove(site)
    }

    enum class ImportExport {
        Import,
        Export
    }

    interface ImportExportCallbacks {
        fun onSuccess(importExport: ImportExport)
        fun onNothingToImportExport(importExport: ImportExport)
        fun onError(error: Throwable, importExport: ImportExport)
    }

    class DowngradeNotSupportedException(message: String) : Exception(message)

    companion object {
        private const val TAG = "ImportExportRepository"

        // Don't forget to change this when changing any of the Export models.
        // Also, don't forget to handle the change in the onUpgrade or onDowngrade methods
        const val CURRENT_EXPORT_SETTINGS_VERSION = 6
    }
}
